### ISSUE TITLE
[Controller] adding Actions as an invokable objects

### ISSUE THREAD

**Issue by wtorsi** [symfony/symfony#38511]
Hi all

I want to propose some changes to `AbstractController` and it's behaviour, i suppose it can increase the overall performance of controllers.

**The goal**
1. The main idea is to make a controller as slim as possible, therefore I suggest adding responses as objects, which looks as a `standalone controllers` and remove `subscribedServises` from the root controller at all (if it is not needed).
2. Achieve standardized responses (mostly for APIs), in this request I called them `actions`.

The request validation services as `Forms` in my opinion should still be left in the controller action as is. Because their purpose is to validate the request, and return mapped data, and not do action.

**The problem**
Imagine the situations  
- you need to perform some actions depending on the request and return full or slim response for `xmlHttpRequest` or not at the same time.
- you work on the API where each kind of object should be formatted.

In both cases your code of actions becomes bigger, you will have a lot of dependencies which should be injected and passed to deeper methods. 
To eliminate those dependencies somebody can use other injected services as a dependency to build and return a response object.

**The idea**
My idea (thanks to autowiring) pass only an API standardized action object as a dependency, and `__invoke` it with required parameters or propose a new feature. 

**Proposal**
Examples which are described after proposal can be optimised as proposing a new method to `AbstractController`, but in this case all Actions should be public or injected.

```php
class Controller extends AbstractVeryLimitedController
{
    #[Route('/path', name: 'action')]
    public function someAction(Request $request): Response
    {
        // RequestValidation -> FormData
        $this->response(SomeAction::class, $params);
    }
}

// or maybe more widely, thanks to Messenger configuration

class Controller extends AbstractVeryLimitedController
{
    #[Route('/path', name: 'action')]
    public function someAction(Request $request): IterableResponse or StackedResponse
    {
        // RequestValidation -> FormData
        yield SomeAction::class => $someParams;
        yield AnotherAction::class => $anotherParams;
    }
}
```

Below are my not optimised, but working examples.

```php
// before
<?php declare(strict_types=1);

namespace Survey\Controller;

use Survey\Entity\SurveyResult;
use Survey\Form\Data\AbstractQuestionDto;
use Survey\Form\FormFactoryProvider;
use Survey\Manager\AnswerManager;
use Survey\Manager\SurveyResultManager;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

/**
 * @Route("/question", name="question_")
 */
class QuestionController extends \AbstractController
{
    /**
     * @Route("/{surveyResult}", name="view", methods={"GET"})
     */
    public function view(SurveyResult $surveyResult, SurveyResultManager $manager, FormFactoryProvider $formFactoryProvider) : Response
    {
        return $this->createQuestionResponse($surveyResult, $manager, $formFactoryProvider);
    }

    /**
     * @Route("/{surveyResult}/answer", name="answer", methods={"POST"})
     */
    public function doAnswer(SurveyResult $surveyResult, SurveyResultManager $surveyResultManager, AnswerManager $answerManager, FormFactoryProvider $formFactoryProvider) : Response
    {
        $survey = $surveyResult->getSurvey();
        $factory = $formFactoryProvider->getFactory($survey);
        $form = $factory->getForm($surveyResult);

        return $this->handleFormCall($form, function (AbstractQuestionDto $dto) use ($surveyResult, $answerManager, $surveyResultManager, $formFactoryProvider) {
            $answerManager->add($surveyResult, $dto);

            return $this->createQuestionResponse($surveyResult, $surveyResultManager, $formFactoryProvider);
        });
    }

    private function createQuestionForm(SurveyResult $surveyResult, FormFactoryProvider $formFactoryProvider) : FormInterface
    {
        $survey = $surveyResult->getSurvey();
        $factory = $formFactoryProvider->getFactory($survey);
        $data = $factory->getData($surveyResult);

        return $factory->getForm($surveyResult, $data, [
            'action' => $this->generateUrl('question_answer', [
                'surveyResult' => $surveyResult->getId(),
            ]),
        ]);
    }

    private function close(SurveyResultManager $manager, SurveyResult $surveyResult) : Response
    {
        $manager->close($surveyResult);
        $request = $this->container->get('request_stack')->getCurrentRequest();

        $callable = $request->isXmlHttpRequest() ? [$this, 'createRedirectToRouteResponse'] : [$this, 'redirectToRoute'];

        return \call_user_func($callable, 'survey_view', [
            'result' => $surveyResult->getResult()->getId(),
        ]);
    }

    private function createQuestionResponse(SurveyResult $result, SurveyResultManager $manager, FormFactoryProvider $formFactoryProvider) : Response
    {
        if ($result->isCompleted()) {
            return $this->close($manager, $result);
        }

        $form = $this->createQuestionForm($result, $formFactoryProvider);
        $request = $this->container->get('request_stack')->getCurrentRequest();

        if ($request->isXmlHttpRequest()) {
            return $this->createSuccessHtmlResponse('survey/question/_form.html.twig', [
                'form' => $form->createView(),
            ]);
        }

        return $this->render('survey/question/answer.html.twig', [
            'form' => $form->createView(),
        ]);
    }
}
```

```php
/** 
* after
* you may not need ContainerAwareInterface, ServiceSubscriberInterface 
*/ 
<?php declare(strict_types=1);

namespace Survey\Controller;

use Dev\FormBundle\FormTrait;
use Survey\Controller\Action\QuestionAction;
use Survey\Entity\SurveyResult;
use Survey\Form\Data\AbstractQuestionDto;
use Survey\Form\FormFactoryProvider;
use Survey\Manager\AnswerManager;
use Symfony\Component\DependencyInjection\ContainerAwareInterface;
use Symfony\Component\DependencyInjection\ContainerAwareTrait;
use Symfony\Component\Form\FormFactoryInterface;
use Symfony\Component\HttpFoundation\RequestStack;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Contracts\Service\ServiceSubscriberInterface;

/**
 * @Route("/question", name="question_")
 */
class QuestionController implements ContainerAwareInterface, ServiceSubscriberInterface
{
    use FormTrait;
    use ContainerAwareTrait;

    public static function getSubscribedServices()
    {
        return [
            'request_stack' => RequestStack::class,
            'form.factory' => FormFactoryInterface::class,
        ];
    }

    /**
     * @Route("/{surveyResult}", name="view", methods={"GET"})
     */
    public function view(SurveyResult $surveyResult, QuestionAction $action): Response
    {
        return $action($surveyResult);
    }

    /**
     * @Route("/{surveyResult}/answer", name="answer", methods={"POST"})
     */
    public function doAnswer(SurveyResult $surveyResult, AnswerManager $answerManager, FormFactoryProvider $formFactoryProvider, QuestionAction $action): Response
    {
        $survey = $surveyResult->getSurvey();
        $factory = $formFactoryProvider->getFactory($survey);
        $form = $factory->getForm($surveyResult);

        return $this->handleFormCall($form, function (AbstractQuestionDto $dto) use ($answerManager, $surveyResult, $action) {
            $answerManager->add($surveyResult, $dto);

            return $action($surveyResult);
        });
    }
}

<?php
declare(strict_types=1);

namespace Survey\Controller\Action;

use Survey\Entity\SurveyResult;
use Survey\Form\FormFactoryProvider;
use Survey\Manager\SurveyResultManager;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\HttpFoundation\Response;

class QuestionAction extends \Symfony\Bundle\FrameworkBundle\Controller\AbstractAction
{
    private SurveyResultManager $manager;
    private FormFactoryProvider $formFactoryProvider;

    public function __construct(SurveyResultManager $manager, FormFactoryProvider $formFactoryProvider)
    {
        $this->manager = $manager;
        $this->formFactoryProvider = $formFactoryProvider;
    }

    public function __invoke(SurveyResult $result): Response
    {
        if ($result->isCompleted()) {
            return $this->close($result);
        }

        $form = $this->createQuestionForm($result);
        $request = $this->container->get('request_stack')->getCurrentRequest();

        $params = [
            'form' => $form->createView(),
        ];

        if ($request->isXmlHttpRequest()) {
            return $this->createSuccessHtmlResponse('survey/question/_form.html.twig', $params);
        }

        return $this->render('survey/question/answer.html.twig', $params);
    }

    private function close(SurveyResult $surveyResult): Response
    {
        $this->manager->close($surveyResult);
        $request = $this->container->get('request_stack')->getCurrentRequest();

        $callable = $request->isXmlHttpRequest()
            ? [$this, 'createRedirectToRouteResponse']
            : [$this, 'redirectToRoute'];

        return \call_user_func($callable, 'survey_view', [
            'result' => $surveyResult->getResult()->getId(),
        ]);
    }

    private function createQuestionForm(SurveyResult $surveyResult): FormInterface
    {
        $survey = $surveyResult->getSurvey();
        $factory = $this->formFactoryProvider->getFactory($survey);
        $data = $factory->getData($surveyResult);

        return $factory->getForm($surveyResult, $data, [
            'action' => $this->generateUrl('question_answer', [
                'surveyResult' => $surveyResult->getId(),
            ]),
        ]);
    }
}

```

**Comment by pizzaminded** [comment]
IMHO Your problem can be solved by using some DI and [Action-Domain-Responder](https://github.com/pmjones/adr) pattern:


Symfony supports invokable controllers, so there is no need to use controller as a "middleware" for actions. You can define invokable class as a action directly:

````php
<?php

namespace App\Action;

class ViewSomethingAction {

    /**
     * @Route("/view", name="view", methods={"GET"})
     */
    public function __invoke(Request $request, SomeService $service) {
        //run some business logic
        //wrap them in response and return them
    }
}

````
It would be autowired and executed like any other action, no need to wrap them in additional non-invokable classes/methods. 



For solving  "I need to modify my response depending on the request" problem, i would look for ADR:


````php
<?php


class DoSomethingAction
{
    /**
     * this is your controller (A, or Action in ADR Context)
     * @Route("/view", name="view", methods={"GET"})
     */
    public function __invoke(DoSomethingResponder $responder, Request $request)
    {
        //do business logic
        //delegate response creation to responder object
        return $responder($request, $domainObjects);
    }

}

class DoSomethingResponder {

    protected $twig;
    public function __construct(\Twig\Environment $twig)
    {
        $this->twig = $twig;
    }
    
    public function __invoke(Request $request, array $domainObjects): Response
    {
        //based on things you passed here in action, decide which response should be returned

        if($shouldUseSlimResponse) {
            return new Response($this->twig->render('partial.html.twig', $params));
        } 
        
        return new Response($this->twig->render('full.html.twig', $params));
        
        

    }
}

````
In this case, by using autowiring and dependency injection, you are handling only business logic in your actions and delegate creating a response to ``DoSomethingResponder`` object.  No need to juggle domain-related and response-building-related dependencies within single method as both concerns are separated from themselves.

**Comment by Aerendir** [comment]
I think you should never have private methods in controllers: if a method is private, then maybe you need to refactor it in some services and inject the services instead of writing the code in the controller directly.

**Comment by dbrumann** [comment]
As @pizzaminded (nice username btw 😄 ) mentions, you are already quite free in how you want to create your controller. You do not have to use the AbstractController provided by Symfony and pretty much anything callable can be a controller for Symfony.

I doubt that you will generally get better performance from your approach. Symfony employs lazy loading for the services in the controller, so as long as you don't use a service there should be pretty much no penalty. Most performance issues I have encountered in Symfony apps are in the service layer, database calls, file i/o, etc. That's why I think advocating this change to all Symfony users is probably not a good idea, considering how much work it would be to make this change in existing apps with what I expect to be little to no performance gains. The good news is, that you can use this in your projects already and provide the functionality, e.g. as a bundle, and optimize this to your needs.

tl;dr I vote to close this as this is not something that needs to be addressed in the Symfony core and can isntead be solved in user land.

**Comment by fabpot** [comment]
Closing as explained.
